File: app/src/main/java/jp/bellware/echo/filter/BasisStats.java
package jp.bellware.echo.filter;

import jp.bellware.echo.data.StaticQueue;

/**
 * 一定の窓内における基本統計
 */
public class BasisStats {

    private StaticQueue<Float> queue;
    /**
     * 窓サイズ
     * @param window
     */
    public BasisStats(int window){
        this.queue = new StaticQueue<>(window);
    }


    public void filter(float s){
        this.queue.add(s);
    }

    private float average;

    private float max;

    private float sum;

    /**
     * 計算する
     */
    public void calculate(){
        max = Float.MIN_VALUE;
        sum = 0;
        for(int i = 0;i < queue.size();++i){
            float v = queue.get(i);
            sum += v;
            if(v > max)
                max = v;
        }
        average = sum/queue.getCapacity();
    }

    public float getAverage() {
        return average;
    }

    public float getMax() {
        return max;
    }

    public float getSum() {
        return sum;
    }

    public void reset(){
        queue.clear();
    }
}

File: app/src/main/java/jp/bellware/echo/filter/BasisStats.kt
package jp.bellware.echo.filter

/**
 * 一定の窓内における基本統計
 * @param window 窓サイズ
 */
class BasisStats(window: Int) {

    /**
     * 音声キュー
     */
    private val queue: StaticQueue<Float> = StaticQueue(window, 0f)

    /**
     * 平均
     */
    var average = 0f
        private set

    /**
     * 最大
     */
    var max = 0f
        private set

    /**
     * 合計
     */
    var sum = 0f
        private set


    /**
     * 音声パケットを追加する
     */
    fun add(s: Float) {
        this.queue.add(s)
    }

    /**
     * 計算する
     */
    fun calculate() {
        max = java.lang.Float.MIN_VALUE
        sum = 0f
        for (i in 0 until queue.size) {
            val v = queue[i]
            sum += v
            if (v > max)
                max = v
        }
        average = sum / queue.capacity
    }

    /**
     * リセットする
     */
    fun reset() {
        queue.clear()
    }
}
